### PR TITLE
VORTEX Loss

### PR DESCRIPTION
**Pull request by Andrewwango**
### Checks to be done before submitting your PR
- [ ] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: deepinv/loss/augmentation.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+from deepinv.physics.mri import MRI
+from deepinv.physics.forward import LinearPhysics
+from deepinv.physics.inpainting import Inpainting
+from deepinv.transform.base import Transform
+from deepinv.transform.rotate import Rotate
+from deepinv.transform.shift import Shift
+from deepinv.transform.augmentation import RandomPhaseError, RandomNoise
+
+
+class VORTEXLoss(Loss):
```
Comment (tachella) [inline deepinv/loss/augmentation.py:20]: I would call this DACLoss or DataAugmentationConsistencyLoss, since the idea is much more general than the VORTEX paper. We could anyway mention VORTEX in the docstring.

File: deepinv/loss/augmentation.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    r"""Measurement data augmentation loss.
+
+    Performs data augmentation in measurement domain as proposed by
+    `VORTEX: Physics-Driven Data Augmentations Using Consistency Training for Robust Accelerated MRI Reconstruction <https://arxiv.org/abs/2111.02549>`_.
+
+    The loss is defined as follows:
+
+    :math:`\mathcal{L}(T_e\inverse{y,A},\inverse{T_i y,A T_e^{-1}})`
+
+    where :math:`T_1` is a random :class:`deepinv.transform.Transform` defined in k-space and
```
Comment (tachella) [inline deepinv/loss/augmentation.py:30]:
```suggestion
    where :math:`T_e` is a random :class:`deepinv.transform.Transform` defined in image space (that the model is supposed to be equivariant to) and
```

File: deepinv/loss/augmentation.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    :math:`T_2` is a random :class:`deepinv.transform.Transform` defined in image space.
```
Comment (tachella) [inline deepinv/loss/augmentation.py:31]:
```suggestion
    :math:`T_i` is a random :class:`deepinv.transform.Transform` defined in measurement space (that the model is supposed to be invariant to).
```

File: deepinv/loss/augmentation.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    By default, for :math:`T_1` we add random noise :class:`deepinv.transform.RandomNoise` and random phase error :class:`deepinv.transform.RandomPhaseError`.
```
Comment (tachella) [inline deepinv/loss/augmentation.py:32]:
```suggestion
    By default, for :math:`T_i` we add random noise :class:`deepinv.transform.RandomNoise`.
```

I would remove the default random phase error default since it only holds for MRI

File: deepinv/loss/augmentation.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    By default, for :math:`T_2` we use random shift :class:`deepinv.transform.Shift` and random rotates :class:`deepinv.transform.Rotate`.
```
Comment (tachella) [inline deepinv/loss/augmentation.py:33]:
```suggestion
    By default, for :math:`T_e` we use random shift :class:`deepinv.transform.Shift` and random rotates :class:`deepinv.transform.Rotate`.
```

File: deepinv/loss/augmentation.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    .. note::
+
+        See :ref:`transform` for a guide on all available transforms, and how to compose them. For example, you can easily
+        compose further transforms such as  ``Rotate(rng=rng, multiples=90) | Scale(factors=[0.75, 1.25], rng=rng) | Reflect(rng=rng)``.
+
+    .. note::
+
+        For now, this loss is only available for MRI and Inpainting problems, but it is easily generalisable to other problems.
+
+    :param deepinv.transform.Transform T_1: k-space transform.
```
Comment (tachella) [inline deepinv/loss/augmentation.py:44]:
```suggestion
    :param deepinv.transform.Transform T_i: measurement domain transform for which the model should be invariant.
```

File: deepinv/loss/augmentation.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    :param deepinv.transform.Transform T_2: image transform.
```
Comment (tachella) [inline deepinv/loss/augmentation.py:45]:
```suggestion
    :param deepinv.transform.Transform T_e: image domain transform for which the model should be equivariant.
```

File: deepinv/loss/augmentation.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            )
+
+        def A_adjoint(self, y, *args, **kwargs):
+            return self.transform(
+                super().A_adjoint(y, *args, **kwargs), **self.transform_params
+            )
+
+    def forward(self, x_net: Tensor, y: Tensor, physics: MRI, model, **kwargs):
+        r"""
+        VORTEX loss forward pass.
```
Comment (tachella) [inline deepinv/loss/augmentation.py:116]: change name

File: deepinv/loss/augmentation.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        :return: (:class:`torch.Tensor`) loss, the tensor size might be (1,) or (batch size,).
+        """
+        if self.no_grad:
+            x_net = x_net.detach()
+
+        # Sample image transform
+        e_params = self.T_2.get_params(x_net)
+
+        # Augment input
+        x_aug = self.T_2(physics.A_adjoint(self.T_1(y)), **e_params)
```
Comment (tachella) [inline deepinv/loss/augmentation.py:132]: I'm not sure why you are adding T_2 here

**Review by tachella — CHANGES_REQUESTED** [review]
thanks for the PR

- I think we should fix the naming of the transforms (the docs mix T_1, T_2 with T_i T_e), keeping only one option
- I would change the name of the loss since this technique is common beyond vortex (it is called data augmentation consistency)
- I would remove the MRI-specific default transforms to make it more generally applicable
- I have a doubt regading the implementation

File: deepinv/loss/augmentation.py (hunk 9 above)
Comment (Andrewwango) [reply deepinv/loss/augmentation.py:132]: This is from the [paper](https://arxiv.org/pdf/2111.02549) (Fig. 1)?

File: deepinv/loss/augmentation.py (hunk 1 above)
Comment (Andrewwango) [reply deepinv/loss/augmentation.py:20]: done

File: deepinv/loss/augmentation.py (hunk 8 above)
Comment (Andrewwango) [reply deepinv/loss/augmentation.py:116]: done

File: deepinv/loss/augmentation.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        # Transform physics
+        physics2 = deepcopy(physics)
+        A, A_adjoint, A_dagger = physics2.A, physics2.A_adjoint, physics2.A_dagger
+        physics2.A = lambda x, *args, **kwargs: A(
+            self.T_e.inverse(x, **e_params), *args, **kwargs
+        )
+        physics2.A_adjoint = lambda y, *args, **kwargs: self.T_e(
+            A_adjoint(y, *args, **kwargs), **e_params
+        )
+        physics2.A_dagger = lambda y, *args, **kwargs: self.T_e(
```
Comment (tachella) [inline deepinv/loss/augmentation.py:103]: this only works for linear physics

I would prefer to compose physics as is done [here](https://github.com/deepinv/deepinv/blob/8189cbc8a75888fce1759dbe2564ecb2601f6c60/deepinv/transform/base.py#L297), where we first define a linear physics associated with the transform and then we just concatenate with the base physics.

File: deepinv/transform/augmentation.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        return torch.cat([n(y) * mask for n in noise_model])
+
+    def inverse(self, *args, **kwargs):
+        raise ValueError("Noise transform is not invertible.")
+
+
+class RandomPhaseError(Transform):
+    """Random phase error transform.
+
+    This transform adds a phase error to k-space using:
```
Comment (tachella) [inline deepinv/transform/augmentation.py:68]:
```suggestion
    This transform is specific to MRI problems, and adds a phase error to k-space using:
```

File: deepinv/transform/augmentation.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+    :math:`Ty=\exp(-i\phi_k)y` where :math:`\phi_k=\pi\alpha s_e` if :math:`k` is an even index,
+    or :math:`\phi_k=\pi\alpha s_o` if odd, and where :math:`\alpha` is a scale parameter,
+    and :math:`s_o,s_e\sim U(-1,1)`.
+
+    This transform is reproducible: for given param dict `se, so`, the transform is deterministic.
+
+    :param int, tuple[int, int] scale: scale parameter or range to pick randomly.
```
Comment (tachella) [inline deepinv/transform/augmentation.py:76]:
```suggestion
    :param int, tuple[int, int] scale: scale parameters :math:`s_e` and :math:`s_o` or range to pick randomly.
```

**Review by tachella — CHANGES_REQUESTED** [review]

File: deepinv/loss/augmentation.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        T_e: Transform = None,
+        metric: Union[Metric, nn.Module] = torch.nn.MSELoss(),
+        no_grad: bool = True,
+        rng: torch.Generator = None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.metric = metric
+        self.T_i = T_i if T_i is not None else RandomNoise(rng=rng)
```
[the remark this answers is not in the document]
Comment (Andrewwango) [reply deepinv/loss/augmentation.py:66]: done

File: deepinv/transform/augmentation.py (hunk 11 above)
Comment (Andrewwango) [reply deepinv/transform/augmentation.py:68]: done

File: deepinv/loss/augmentation.py (hunk 10 above)
Comment (Andrewwango) [reply deepinv/loss/augmentation.py:101]: I agree, that it would be nice to do:
```python
        # Transform physics
        physics2 = physics * LinearPhysics(
            A=lambda x: self.T_e.inverse(x, **e_params),
            A_adjoint=lambda x: self.T_e(x, **e_params),
        )
```
But I just tried this, and the results are very different. Something strange is going on.

I would propose leaving it as is for now, and when your equivaraint PR is ready for review, I can try changing this again.

Comment (Andrewwango) [reply deepinv/loss/augmentation.py:101]: Added issue here https://github.com/deepinv/deepinv/issues/489

**Review by tachella — APPROVED** [review]
